testctrl/svc/orch/controller.go

```diff
@@ -70,11 +68,44 @@ func (c *Controller) Start() error {
 		return fmt.Errorf("controller start failed when setting up queue: %v", err)
 	}
 
-	if err := c.startWatcher(); err != nil {
+	if err := c.watcher.Start(c.clientset.CoreV1().Pods(corev1.NamespaceDefault)); err != nil {
 		return fmt.Errorf("controller start failed when starting watcher: %v", err)
 	}
 
-	c.startExecutors()
+	go func() {
+		for {
+			if c.quit {
+				return
+			}
+
+			if c.activeCount >= executorCount {
+				continue // loop until some executors are released
+			}
+
+		retryDequeue:
+			session := c.queue.Dequeue()
+			if session == nil {
+				goto retryDequeue // loop until machines are available
```

Comment:
I believe this is the busy wait I was referring to (unless there are others). Nothing between lines 85 and 88 yields the goroutine (as far as I can tell), so this is eating a bunch of CPU cycles unnecessarily. I haven't looked all the way through the stack to see whether the k8s client is using [the watch API](https://kubernetes.io/docs/reference/using-api/api-concepts/#efficient-detection-of-changes) or not. If it is, then the ideal should be to wake up your stack only when an event is sent from the `kube-apiserver` (modulo spurious wakeups).

Reply:
I figured you were referring to this logic which waits on kubernetes: https://github.com/codeblooded/grpc/pull/75/files#diff-71ccf3b54a1ea98d2ea507f6611ce264R90 which has been fixed.

In the latest rebase, I have added a timeout to space this out. This code is dependent on a custom queue, which only dequeues if it estimates there are enough available machines. 
The bookkeeping is based on the number of machines at startup. It does not stay in contact with kubernetes, but I would like to change that eventually.

Either way, I'll find a way to improve this. In the meantime, it should address the CPU utilization concerns.